release.py

```diff
@@ -0,0 +1,848 @@
+        res += full_changelog
+
+    return res
+
+
+def print_release_notes(notes: str):
+    """Print formatted release notes.
+
+    So that user can review them and use them to write the title and summary.
+    """
+    print('=' * 80)
+    print('Formatted release notes:')
+    print('=' * 80)
+    print(notes)
+    print('=' * 80)
+
+
+def input_title_and_summary(release: github.GitRelease.GitRelease) -> tuple[str, str]:
+    """Ask user to input the release title and summary."""
+    logger.info(f'The automatically generated title is: {release.title}')
+    title = input('Enter release title, press Enter to keep the auto-generated title:\n> ').strip()
+    if not title:
+        title = release.title
+
+    print("\nEnter release summary (multi-line supported; type '.' on a new line to finish):")
+
+    lines: list[str] = []
+    while True:
+        line = input()
+        if line.strip() == '.':
+            break
+        lines.append(line)
+
+    summary = '\n'.join(lines).strip()
+    summary = summary.lstrip('\n')
+    summary = summary.rstrip('\n') + '\n'
+
+    return title, summary
+
+
+def update_draft_release(release: github.GitRelease.GitRelease, title: str, notes: str):
+    """Update the release with the provided title and notes."""
+    max_retries = 3
+    for attempt in range(1, max_retries + 1):
+        try:
+            release.update_release(name=title, message=notes, draft=True)
+            break
+        except Exception as e:
+            logger.error(f'Attempt {attempt} failed to update release: {e}')
+        if attempt == max_retries:
+            raise
+        time.sleep(1)
+    logger.info('Release title and notes updated.')
+
+
+def format_changes(categories: dict[str, list[tuple[str, str]]], tag: str) -> str:
+    """Format for CHANGES.md.
+
+    The header is formatted as a top-level heading with the tag and date.
+    The content is a Markdown formatted string with sections for each commit type.
+    Each item is formatted as a bullet point with the description and PR number in parentheses.
+    """
+    today = datetime.now().strftime('%d %B %Y')
+    res = f'# {tag} - {today}\n\n'
+
+    for commit_type, items in categories.items():
+        if items:
+            res += f'## {commit_type_to_category(commit_type)}\n\n'
+
+            for description, pr_link in items:
+                pr_num = '?'
+                match = re.match(r'https?://[^ ]+/pull/(\d+)', pr_link)
+                if match:
+                    pr_num = match.group(1)
+                res += f'* {description} (#{pr_num})\n'
+
+            res += '\n'
+
+    return res
+
+
+def update_changes_file(changes: str, file: str):
+    """Update the changes file with new release notes."""
+    existing_content = ''
+    if os.path.exists(file):
+        with open(file) as f:
+            existing_content = f.read()
+
+    new_content = changes + existing_content
+
+    with open(file, 'w') as f:
+        f.write(new_content)
+
```

Comment:
We should probably use `pathlib` for path stuff. Then this could be, e.g.:
```suggestion
    existing_content = path.read_text() if path.exists() else ''
    path.write_text(changes + existing_content)
```